### DISCUSSION TITLE
plot_decision_regions filler_feature_values filler_feature_ranges

### DISCUSSION THREAD

**Discussion by kurikabocya** [rasbt/mlxtend#874] [Q&A]
I can't understand plot_decison_regions's filler_feature_values and filler_feature_ranges
Here the filler_feature_values says that the featuer No.3's value is 1.5
Here the filler_feature_ranges says that the featuer No.3's width is 0.75
But what is that 1.5 and 0.75 ?

<Code>
X, y = datasets.make_blobs(n_samples=600, n_features=3,
                           centers=[[2, 2, -2],[-2, -2, 2]],
                           cluster_std=[2, 2], random_state=2)

# Training a classifier
svm = SVC(gamma='auto')
svm.fit(X, y)

# Decision region for feature 3 = 1.5
value = 1.5
# Plot training sample with feature 3 = 1.5 +/- 0.75
width = 0.75

plot_decision_regions(X, y, clf=svm,
                      feature_index=[0,1],
                      filler_feature_values={2: value},
                      filler_feature_ranges={2: width},
                      legend=2, ax=axes[0])

**Comment by rasbt** [comment]
You can only plot 2 features at the time in a 2D scatterplot, so how do you deal with the other features? If you have 3 features, the feature space would be a cube, and then choosing a filler feature like 1.5 for that feature would be like taking a slice. So, it is essentially making the 2D scatterplot for features 1 and 2 while keeping feature 3 constantly at 1.5. I hope that makes sense?

**Reply by kurikabocya** [reply]
Thank you.
Again the above example.
plot_decision_regions(X, y, clf=svm, feature_index=[0,1], filler_feature_values={2: value}, filler_feature_ranges={2: width}, legend=2, ax=axes[0])

This means the feature No.3 is displayed when No.3 value is between 2 - 2 to 2 + 2 or 0 to 4.
If X is
[0,1,2]
[2,1,3]
[-2,4,200]
[-4,2,-1],
only
[0,1,2]
[2,1,3]
are displayed.

**Comment by rasbt** [comment]
Good question. It's definitely a bit tricky to understand this function, and the documentation could definitely be improved with some better illustration or example. So what the `filler_feature_values` does is selecting all data points where the feature value is exactly that particular value. If you set the `filler_feature_ranges` in addition to that, then it considers data points where the value is in the range [value - width, value + width]. So, I think your understanding

> This means the feature No.3 is displayed when No.3 value is between 2 - 2 to 2 + 2 or 0 to 4.

is correct!
